intel-oneapi-compilers: add  to LD_LIBRARY_PATH so that it finds libimf.so

frankwillmore opened the pull request:
Amendment to #20686, adds path to LD_LIBRARY_PATH so that it finds libimf.so. @rscohn2 will libimf.so still be required or did this turn out to be a compiler bug after all? 

@scheibelp

frankwillmore committed:
add  to LD_LIBRARY_PATH so that it finds libimf.so

frankwillmore committed:
add  to LD_LIBRARY_PATH so that it finds libimf.so
and cleanup PEP8

frankwillmore committed:
add  to LD_LIBRARY_PATH so that it finds libimf.so
and cleanup PEP8

frankwillmore committed:
add  to LD_LIBRARY_PATH so that it finds libimf.so
and cleanup PEP8

frankwillmore committed:
add  to LD_LIBRARY_PATH so that it finds libimf.so
and cleanup PEP8

frankwillmore committed:
add  to LD_LIBRARY_PATH so that it finds libimf.so

frankwillmore committed:
add  to LD_LIBRARY_PATH so that it finds libimf.so

frankwillmore committed:
PEP8 cleanup

rscohn2 committed:
refactor path logic

rscohn2 committed:
refactor path logic

rscohn2 committed:
refactor path logic

rscohn2 committed:
refactor path logic

rscohn2 committed:
refactor path logic

rscohn2 committed:
refactor path logic

frankwillmore committed:
Merge pull request #4 from rscohn2/libimf

refactor path logic

frankwillmore commented:
incorporating @rscohn2 's refactor

scheibelp commented:
Note that https://github.com/spack/spack/pull/20411 updated the Spack `oneapi` compiler definition to list `libimf` as a required dependency, so when building packages with `icx` in Spack, they should add RPATHs that include the directory with `libimf`.

Is this for generating a module file to support using a Spack-installed OneAPI compiler outside of Spack? If so then this could be useful (e.g. making sure the directory including `libimf.so` is part of `LD_LIBRARY_PATH`. Is that the only library that's needed? If so then `_ld_library_path` may be adding more directories than required.

Another way for me to approach this is to check if you are still having issues using the OneAPI compilers with Spack and whether that is in the context of:

* externally-installed OneAPI compiler
* Spack-installed OneAPI compiler

and also in the latter case if you are

* building in Spack or
* building outside of Spack

rscohn2 commented:
What follows assumes the intention of this change is for generating a module file to satisfy compile-time dependencies of the compilers.

Intel does not define which libraries are needed during compile-time. We just say to source setvars.sh script, which will also add directories that are only needed at runtime. My thought in this change was to make the behavior the same as for users who use the setvars.sh script. Another approach is to only add the directories we know are needed. I think that is only linux/compiler/lib/intel64_lin. I can change it to that if there is agreement.

I also want to point out that the use of _ld_library_path in install is also to satisfy compile-time dependencies so it should be consistent.

frankwillmore commented:
To summarize the outstanding issue:
- When spack is building packages whose configure process requires building and running intermediate executables with the OneAPI compilers, these built executables are not able to find libimf.so. 
- The fix in this pull requests solve this problem for *running* the OneAPI compilers by including explicit rpaths, but does not address the need of the built executables. 
- per @rscohn2 's comment, Intel's position has been that the setvars.sh script should be sourced before running compilation, and for outside-of-spack builds, doing so addresses both issues. 
- By first loading the generated oneapi module before running builds inside or outside of spack, both issues are similarly addressed. 
- To get the same effect, the compiler stanza which is generated by detection would need to be edited by hand to either include path/to/libimf.so in its extra_rpaths section, or to include the requirement of the generated oneapi module. 
- per @becker33 's comment in slack (please confirm or clarify), it may be possible to override implicit_rpaths() in lib/spack/spack/compiler.py in the oneapi subclass to force needed rpaths into generated executables. 

The final point (if my understanding of it is correct) would be the most robust solution, but would take some effort to implement. There is no example from other compilers, as this issue is specific to intel. 
In lieu of that, and in preparation for the E4S spack hack-a-thon tomorrow, I'm recommending that the generated oneapi module be manually added to the generated compiler stanza before any builds, e.g.: 
```
- compiler:
    spec: oneapi@2021.1
    paths:
      cc: /soft/spack/opt/spack/linux-opensuse_leap15-cascadelake/gcc-10.1.0/intel-oneapi-compilers-2021.1.0-wgzjfktx5hsx75gghtd6k2qxad5vnot2/compiler/latest/linux/bin/icx
      cxx: /soft/spack/opt/spack/linux-opensuse_leap15-cascadelake/gcc-10.1.0/intel-oneapi-compilers-2021.1.0-wgzjfktx5hsx75gghtd6k2qxad5vnot2/compiler/latest/linux/bin/icpx
      f77: /soft/spack/opt/spack/linux-opensuse_leap15-cascadelake/gcc-10.1.0/intel-oneapi-compilers-2021.1.0-wgzjfktx5hsx75gghtd6k2qxad5vnot2/compiler/latest/linux/bin/ifx
      fc: /soft/spack/opt/spack/linux-opensuse_leap15-cascadelake/gcc-10.1.0/intel-oneapi-compilers-2021.1.0-wgzjfktx5hsx75gghtd6k2qxad5vnot2/compiler/latest/linux/bin/ifx
    flags: {}
    operating_system: opensuse_leap15
    target: x86_64
    modules: 
    - /soft/spack/share/spack/modules/linux-opensuse_leap15-cascadelake/intel-oneapi-compilers-2021.1.0-gcc-10.1.0-wgzjfkt
    environment: {}
    extra_rpaths: [] 
```

frankwillmore committed:
added some paths to make compilers and libs discoverable

frankwillmore committed:
added some paths to make compilers and libs discoverable

frankwillmore committed:
Merge branch 'libimf' of https://github.com/frankwillmore/spack into libimf

scheibelp commented:
> When spack is building packages whose configure process requires building and running intermediate executables with the OneAPI compilers, these built executables are not able to find libimf.so.

> The fix in this pull requests solve this problem for running the OneAPI compilers by including explicit rpaths, but does not address the need of the built executables.

When Spack is building packages with `%oneapi` my intent was that it would add these RPATHs (e.g. for `libimf`) to any compiled executable, and I thought I had confirmed as much in https://github.com/spack/spack/pull/20411#issuecomment-750432601, but it sounds like you are seeing a counterexample: could you mention one of the packages that is failing for you?

> per @rscohn2 's comment, Intel's position has been that the setvars.sh script should be sourced before running compilation, and for outside-of-spack builds, doing so addresses both issues.

If this needs to be run whenever using the `%oneapi` compiler, I could look into adding a hook to do that in Spack (or e.g. to apply all the settings it mentions in its module file since you mention that works too).

rscohn2 commented:
I got confirmation that the need to set LD_LIBRARY_PATH to invoke the compiler is likely to go away in the next release. Given that, I think it is ok to explicitly add the directories we need for now. Sourcing the setvars would be better at future proofing, but that is not necessary.

frankwillmore committed:
Merge pull request #6 from spack/develop

adios2: Define custom libs property (#20848)

mic84 committed:
amrex: fix handling of CUDA arch (#20786)

* amrex: fix handling of CUDA arch
* amrex: fix style
* amrex: fix bug
* Update var/spack/repos/builtin/packages/amrex/package.py
* Update var/spack/repos/builtin/packages/amrex/package.py

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

Commit:
ecp-data-vis-sdk: Combine the vis and io SDK packages (#20737)

This better enables the collective set to be deployed togethor satisfying
eachothers dependencies

darmac committed:
r-sf: fix dependency error (#20898)

dtaller committed:
improve documentation for Rocm (hip amd builds) (#20812)

* improve documentation

t-nojiri committed:
astyle: Fix makefile for install parameter (#20899)

frankwillmore committed:
adding paths to LIBRARY_PATH so compiler wrappers will find -lmpi

frankwillmore committed:
adding paths to LIBRARY_PATH so compiler wrappers will find -lmpi

shintaro-iwasaki committed:
llvm-doe: added new package (#20719)

The package contains duplicated code from llvm/package.py,
will supersede solve.

glennpj committed:
r-e1071: added v1.7-4 (#20891)

glennpj committed:
r-diffusionmap: added v1.2.0 (#20881)

glennpj committed:
r-covr: added v3.5.1 (#20868)

glennpj committed:
r-class: added v7.3-17 (#20856)

KineticTheory committed:
py-h5py: HDF5_DIR is needed for ~mpi too (#20905)

For the `~mpi` variant, the environment variable `HDF5_DIR` is still required.  I moved this command out of the `+mpi` conditional.

henrique committed:
py-hovorod: fix typo on variant name in conflicts directive (#20906)

t-nojiri committed:
fujitsu-fftw: Add new package (#20824)

frankwillmore committed:
added vals for CC=icx, CXX=icpx, FC=ifx to generated module

frankwillmore committed:
added vals for CC=icx, CXX=icpx, FC=ifx to generated module

h-denpo committed:
pocl: added v1.6 (#20932)

Made version 1.5 or lower conflicts with a64fx.

adamjstewart committed:
PCL: add new package (#20933)

glennpj committed:
r-rle: new package (#20916)

Common 'base' and 'stats' methods for 'rle' objects, aiming to make it
possible to treat them transparently as vectors.

glennpj committed:
r-ellipsis: added v0.3.1 (#20913)

t-nojiri committed:
libconfig: add build dependency on texinfo (#20930)

glennpj committed:
r-flexmix: add v2.3-17 (#20924)

glennpj committed:
r-fitdistrplus: add v1.1-3 (#20923)

glennpj committed:
r-fit-models: add v0.64 (#20922)

glennpj committed:
r-fields: add v11.6 (#20921)

glennpj committed:
r-fftwtools: add v0.9-9 (#20920)

glennpj committed:
r-farver: add v2.0.3 (#20919)

glennpj committed:
r-expm: add v0.999-6 (#20918)

t-nojiri committed:
cln: add build dependency on texinfo (#20928)

glennpj committed:
r-expint: add v0.1-6 (#20917)

glennpj committed:
r-envstats: add v2.4.0 (#20915)

glennpj committed:
r-energy: add v1.7-7 (#20914)

glennpj committed:
r-ellipse: add v0.4.2 (#20912)

adamjstewart committed:
py-fiscalyear: add v0.3.0 (#20911)

glennpj committed:
r-ecp: add v3.1.3 (#20910)

glennpj committed:
r-plotmo: add v3.6.0 (#20909)

tomgreen66 committed:
Improve gcc detection in llvm. (#20189)

Co-authored-by: Tom Scogland <tom.scogland@gmail.com>
Co-authored-by: Thomas Green <ca-tgreen@gw4a64fxlogin00.head.gw4.metoffice.gov.uk>

bhatele committed:
hatchet: updated urls (#20908)

adamjstewart committed:
py-anuga: add new package (#20782)

darmac committed:
libvips: added v8.10.5 (#20902)

a-saitoh-fj committed:
libzmq: add platform conditions to libbsd dependency (#20893)

glennpj committed:
r-dtw: add v1.22-3 (#20890)

glennpj committed:
r-dt: add v0.17 (#20889)

glennpj committed:
r-dosnow: add v1.0.19 (#20888)

glennpj committed:
add version 1.0.16 to r-doparallel (#20886)

glennpj committed:
add version 1.3.7 to r-domc (#20885)

glennpj committed:
add version 0.9-15 to r-diversitree (#20884)

glennpj committed:
add version 1.3-3 to r-dismo (#20883)

glennpj committed:
add version 0.6.27 to r-digest (#20882)

glennpj committed:
add version 1.5 to r-rngtools (#20887)

glennpj committed:
add version 1.5.8 to r-dicekriging (#20877)

glennpj committed:
add version 1.4.2 to r-httr (#20876)

glennpj committed:
add version   1.28 to r-desolve (#20875)

glennpj committed:
add version   2.2-5 to r-deoptim (#20874)

glennpj committed:
add version   0.2-3 to r-deldir (#20873)

glennpj committed:
add version   1.0.0 to r-crul (#20870)

glennpj committed:
add version   1.1.0.1 to r-crosstalk (#20869)

glennpj committed:
add version   1.0-1 to r-copula (#20867)

glennpj committed:
add version 5.0.2 to r-rcppparallel (#20866)

glennpj committed:
add version   2.0-1 to r-compositions (#20865)

glennpj committed:
add version 0.4.10 to r-rlang (#20796)

glennpj committed:
add version 0.3.6 to r-vctrs (#20878)

mic84 committed:
amrex: add ROCm support (#20809)

glennpj committed:
add version 2.0-0 to r-colorspace (#20864)

glennpj committed:
add version 1.3-1 to r-coin (#20863)

glennpj committed:
add version   0.19-4 to r-coda (#20862)

glennpj committed:
add version 1.3.7 to r-clustergeneration (#20861)

glennpj committed:
add version 0.3-58 to r-clue (#20860)

glennpj committed:
add version 0.7.1 to r-clipr (#20859)

glennpj committed:
add version 2.2.0 to r-cli (#20858)

glennpj committed:
add version 0.4-3 to r-classint (#20857)

glennpj committed:
add version 0.1.2 to r-globaloptions (#20855)

glennpj committed:
add version 2.3-56 to r-chron (#20854)

glennpj committed:
add version 0.4.10 to r-checkpoint (#20853)

glennpj committed:
add version 2.0.0 to r-checkmate (#20852)

glennpj committed:
add version 1.18.1 to r-catools (#20850)

glennpj committed:
add version 1.2.2.2 to r-modelmetrics (#20849)

glennpj committed:
add version 3.0-4 to r-cardata (#20847)

glennpj committed:
add version 1.0.1 to r-caracas (#20846)

glennpj committed:
r-lifecycle: new package at v0.2.0 (#20845)

glennpj committed:
add version 3.0-10 to r-car (#20844)

glennpj committed:
add version 3.4.5 to r-processx (#20843)

glennpj committed:
add version 1.5-12.2 to r-cairo (#20842)

glennpj committed:
add version 0.2.3 to r-cubist (#20841)

glennpj committed:
add version 2.6 to r-rmarkdown (#20838)

glennpj committed:
add version 1.2.1 to r-blob (#20819)

glennpj committed:
add version 4.0.4 to r-bit (#20818)

glennpj committed:
add version 2.4-1 to r-bio3d (#20816)

glennpj committed:
add version 0.4.2.3 to r-bibtex (#20815)

glennpj committed:
add version 3.1-4 to r-bayesm (#20807)

glennpj committed:
add version 1.2.1 to r-backports (#20806)

glennpj committed:
add version 2.0.3 to r-argparse (#20805)

glennpj committed:
add version 5.4-1 to r-ape (#20804)

glennpj committed:
add version 0.8-18 to r-amap (#20803)

glennpj committed:
r-pixmap: added new package (#20795)

Commit:
zoltan: source code location change (#20787)

frankwillmore committed:
fix merge

frankwillmore commented:
The issue has gone away. I was working with several branches and got my wires crossed.  

I'm trying to clean it up a bit, should be close to ready.

frankwillmore committed:
Merge branch 'libimf' of https://github.com/frankwillmore/spack into libimf

frankwillmore committed:
Merge branch 'develop' of https://github.com/spack/spack into develop

frankwillmore committed:
Merge branch 'develop' into libimf

frankwillmore committed:
back out changes to intel-oneapi-mpi, save for separate PR

rscohn2 commented on var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:
```diff
@@ -38,35 +37,45 @@ def __init__(self, spec):
+        env.prepend_path('CPATH', self._join_prefix('include'))
+        env.prepend_path('LIBRARY_PATH', self._join_prefix('lib'))
+        for dir in self._ld_library_path():
+            env.prepend_path('LD_LIBRARY_PATH', self._join_prefix(dir))
```
```suggestion
            env.prepend_path('LD_LIBRARY_PATH', dir)
```

Looks like the unnecessary join_prefix came from me. I was only testing spack-installed compiler which does not need this. I need to add a test with externally installed compiler.

frankwillmore committed:
Update var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py

path is joined in _ld_library_path()

Co-authored-by: Robert Cohn <rscohn2@gmail.com>

frankwillmore commented:
Do we need anything else on this? I'm doing some git housecleaning.

scheibelp commented on var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:
```diff
@@ -38,35 +37,45 @@ def __init__(self, spec):
+            env.prepend_path('LD_LIBRARY_PATH', dir)
+        env.set('CC', 'icx')
+        env.set('CXX', 'icpx')
+        env.set('FC', 'ifx')
```
The functional changes in this PR appear to be:

* Add all entries of `_ld_library_path()` to `LD_LIBRARY_PATH` rather than just `lib/`
* Set CC and other compiler vars

Is the first one necessary? I mentioned in https://github.com/spack/spack/pull/20717#issuecomment-759155895 that I wouldn't expect that to be required: we should automatically be adding RPATHs to the compiled binaries (since `libimf` is in `required_libs` for the `oneapi` compiler class defined in Spack).

scheibelp replied:
EDIT (it was established that this wouldn't work) ~The part that sets CC might be useful as an alternative to setting `I_MPI_CC` in https://github.com/spack/spack/pull/20808/files#r564934758~

scheibelp replied:
`environment_modifications_for_spec` adds `lib/` to `LD_LIBRARY_PATH`, so the choice to make sure the appropriate lib directories are added isn't a big deal. I was curious though if it was strictly necessary. No changes are needed here.

rscohn2 replied:
It looks good.

scheibelp commented:
@frankwillmore the previous message I see here before

> Do we need anything else on this?

Is https://github.com/spack/spack/pull/20717#issuecomment-761243713

> The issue has gone away. I was working with several branches and got my wires crossed.

In relation to https://github.com/spack/spack/pull/20717#discussion_r564927835, I want to check if you're asking whether this should be merged or closed. If you are currently seeing an issue with using `oneapi` in Spack develop then I should look at this more (and sorry for getting confused).

frankwillmore commented:
Merged. The issue I had above was from me losing track of branches. It's working for me now, wouldn't hurt to have another set of eyes go over it

tldahlgren commented:
@rscohn2 @scheibelp Is this PR ready to be merged?

scheibelp commented on var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:
```diff
@@ -38,35 +37,45 @@ def __init__(self, spec):
+        env.prepend_path('LIBRARY_PATH', self._join_prefix('lib'))
+        for dir in self._ld_library_path():
+            env.prepend_path('LD_LIBRARY_PATH', dir)
+        env.set('CC', 'icx')
```
Could this set CC to point to the absolute path? Setting `PATH` should avoid most issues but generally we are explicit anyway (e.g. in the Spack `gcc` package.py).

frankwillmore replied:
I've updated for absolute paths, and verified they are correct in the generated module. Should be good to go.

scheibelp reviewed (changes requested):
Could you set CC/CXX/FC to point to absolute paths?

scheibelp commented:
Other than updating CC etc. this looks good. Sorry for losing track of this.

frankwillmore committed:
set absolute paths to icx,icpx,ifx

frankwillmore committed:
dang close parenthesis

tldahlgren commented:
Thank you for your contribution!
